scipy/cluster/hierarchy.py

```diff
@@ -1395,6 +1400,184 @@ def cut_tree(Z, n_clusters=None, height=None):
     return groups.T
 
 
+def cut_tree_balanced(Z, max_cluster_size):
+    """
+    Given a linkage matrix Z and max cluster size, return a balanced cut tree.
+
+    The function looks recursively along the hierarchical tree, from the root
+    (single cluster gathering all the samples) to the leaves (i.e. the clusters
+    with only one sample), retrieving the biggest possible clusters containing
+    a number of samples lower than a given maximum. If a cluster at a specific
+    tree level contains a number of samples higher than the given maximum, it
+    is ignored and its offspring (smaller) sub-clusters are taken into
+    consideration. If the cluster contains a number of samples lower than the
+    given maximum, it is taken as result and its offspring sub-clusters not
+    further processed.
+
+    Parameters
+    ----------
+    Z : ndarray
+        The linkage matrix resulting from calling `ward` or `linkage`. I.e. it 
+        contains the hierarchical clustering encoded as a linkage matrix.
+    max_cluster_size : int
+        Maximum number of data samples contained within the resulting clusters.
+        Thus, all resulting clusters will contain a number of data samples
+        ``<= max_cluster_size``. Must be >= 1.
+
+    Returns
+    -------
+    cluster_id : ndarray
+        One-dimensional array of integers containing for each input sample its
+        corresponding cluster id. The cluster id is an integer which is higher
+        for deeper tree levels.
+    cluster_level : ndarray
```

Comment:
I left `cluster_level` as a return value just because of one specific reason: it makes clear to the user how similar the returned clusters actually are. Thus, cluster IDs (numbers) only give a hint about how deep are they within the cluster tree (being higher number assigned to deeper levels, as they always increase as the exploration goes deeper). 
However, very different clusters as e.g. `cluster_level=array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1])` and `array([0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1])` could get very similar IDs as `19` and `20` (note that they however represent very different clusters, because they diverge very early, i.e. at second level). 
Thus, through the `cluster_level` the user can get a feeling about the similarity of the resulting clusters, without having to do any further data mining on the raw data. 
Unfortunately, adding this return value to the `cut_tree` method is (to my knowledge) difficult to do, because it follows a different approach. Note that in the case of `cut_tree_balanced`, its computation is simple because of its iterative nature.
Please let me know if you see the return value `cluster_level` as irrelevant to the user. It might be simple omitted (or left as optional) if you prefer so.